lib/mix/lib/mix/tasks/profile.tprof.ex

```diff
@@ -0,0 +1,409 @@
+      (default: `:time`), see [moduledoc](`Mix.Tasks.Profile.Tprof`) for more information
+
+    * `:calls` - filters out any results with a call count lower than this
+    * `:time` - filters out any results that took lower than specified (in µs),
+      `type` needs to be `:time`
+    * `:memory` - filters out any results that used less memory than specified (in words),
+      `type` needs to be `:memory`
+    * `:sort` - sort the results by `:calls` or by the value of `type`
+      (default: the value of `type`)
+    * `:warmup` - if the code should be warmed up before profiling (default: `true`)
+    * `:set_on_spawn` - if newly spawned processes should be measured (default: `true`)
+
+  """
+  @spec profile((-> result), keyword()) :: result when result: any()
+  def profile(fun, opts \\ []) when is_function(fun, 0) do
+    Mix.ensure_application!(:tools)
+    {type, return_value, results} = profile_and_analyse(fun, opts)
+    print_output(type, results)
+    return_value
+  end
+
+  defp profile_and_analyse(fun, opts) do
+    if Keyword.get(opts, :warmup, true) do
+      IO.puts("Warmup...\n")
+      fun.()
+    end
+
+    tprof_module().start()
+    matching = Keyword.get(opts, :matching, {:_, :_, :_})
+    set_on_spawn = Keyword.get(opts, :set_on_spawn, true)
+    type = Keyword.get(opts, :type, :time)
+
+    sort_by =
+      case Keyword.get(opts, :sort) do
+        nil -> :measurement
+        :calls -> :calls
+        ^type -> :measurement
+        other -> Mix.raise("Incompatible sort option `#{other}` with type `#{type}`")
+      end
+
+    tprof_type = to_tprof_type(type)
+
+    {return_value, {^tprof_type, traces}} =
+      tprof_module().profile(fun, %{
+        set_on_spawn: set_on_spawn,
+        pattern: matching,
+        type: tprof_type,
+        report: :return
+      })
+
+    inspected = tprof_module().inspect({tprof_type, traces}, :process, sort_by)
+
+    tprof_module().stop()
+
+    results =
+      inspected
+      |> Enum.map(fn {pid, {^tprof_type, measurement_total, call_results}} ->
+        parsed_calls =
+          call_results
+          |> filter_results(type, opts)
+          |> add_totals(measurement_total)
+
+        {pid, parsed_calls}
+      end)
+
+    {type, return_value, results}
+  end
+
+  defp to_tprof_type(:calls), do: :call_count
+  defp to_tprof_type(:time), do: :call_time
+  defp to_tprof_type(:memory), do: :call_memory
+
+  defp filter_results(call_results, type, opts) do
+    calls_opt = Keyword.get(opts, :calls, 0)
+
+    measurement_opt =
+      get_filter_value!(type, Keyword.get(opts, :time), Keyword.get(opts, :memory))
+
+    Enum.filter(call_results, fn {_module, _fa, count, measurement, _, _} ->
+      count >= calls_opt and measurement >= measurement_opt
+    end)
+  end
+
+  defp get_filter_value!(type, time, _memory) when is_integer(time) and type != :time do
+    Mix.raise("Incompatible use of time option with type `#{type}`")
```

Comment:
We don't use backticks in Elixir error messages. Perhaps #{inspect(type)} is enough?

Reply:
Sounds much better indeed. I needed some kind of visual distinction to keep readability but atom notation feels much more natural.

Reply:
[27e56dc](https://github.com/elixir-lang/elixir/pull/13605/commits/27e56dc01ae5ff6f97c5d175237515bd2226ff4c)